[CK_TILE] Stream-K GEMM Implementation

ecamartins opened the pull request:
## Proposed changes

This PR, made in collaboration with @arai713, contains the first implementation of Stream-K in CK Tile. Our implementation is based on old CK's initial Stream-K implementation (see example/01_gemm/gemm_xdl_streamk.cpp). Stream-K is a GEMM algorithm that helps avoid imbalanced workloads across CUs. When more than one WG is assigned to the same C macro tile, the work is split along the K dimension, where WGs can either atomically add their results to C or select one WG to combine partial results via reduction. Since this is our first implementation of Stream-K in CK Tile, we have opted to implement the atomic strategy with the reduction strategy coming in future work.

For our implementation, this PR makes 3 main additions/alterations:

**1) Updating Universal GEMM's `MakeGemmTensorViews` interface:** The `MakeGemmTensorViews` function has a parameter of type `SplitKBatchOffset`, where only the integer member variable `splitted_k` is used in `MakeGemmTensorViews`. Since Stream-K, in general, is not split-K, we have updated the `MakeGemmTensorViews` interface to require only an integer rather than `SplitKBatchOffset`. This ensures that the same functionality is supported in `MakeGemmTensorViews` while avoiding the constraint of requiring a SplitKBatchOffset.
**2) Adding a `RunGemm` function to the StreamKKernel Class:** The Universal GEMM Kernel’s `RunGemm` function assumes that (a) the kernel uses a `SplitKBatchOffset` and (b) the tile partitioner can statically determine the number of iterations along the K dimension for a given WG (i.e., `num_loop`). These assumptions do not hold for Stream-K; we do not use `SplitKBatchOffset` and an instance method of the tile partitioner is needed to determine `num_loop`. In the Stream-K `RunGemm`, we opted to have `num_loop` as a function parameter to ensure that the main stream-K logic is isolated from `RunGemm`.
**3) Implementing the Stream-K kernel's operator() function:** This function contains the main Stream-K looping logic. In each iteration of the main Stream-K while loop, using the tile partitioner, offsets for the A, B, and C tensors are determined for the current iteration then there is a call to `RunGemm`.

Some other important notes:

- As this is our first implementation, we are focused on correctly implementing the algorithm. Algorithmic optimization will come later.
- ~~There are currently 2 bugs in the `StreamKTilePartitioner` (ticket has already been logged with plans to work on them in upcoming sprints). Such bugs make it such that on certain architectures, certain instances of the stream-K algorithm fail. Thus, for this work in our test suite, we skip tests that have sk_blocks > 0. In other words, we only run cases where each macro tile in C is only worked on by one WG, all other tests are skipped.~~ Update: commit [7328bc7](https://github.com/ROCm/composable_kernel/pull/2781/commits/7328bc73eb927b2037053da41ac816d765d86857) lifts this limitation.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

ecamartins committed:
Change splitk_batch_offset parameter to k_size in UniversalGemmKernel::MakeGemmTensorViews function

Prior to this change, the splitk_batch_offset parameter of
MakeGemmTensorViews had type SplitKBatchOffset. But, the only member
variable of the SplitKBatchOffset class used in the MakeGemmTensorViews
function was splitted_k (an int32_t). The splitted_k value was used as
part of defining the dimensions of the tensor view. That said, for
Stream K, we do not need to use the SplitKBatchOffset class since we are
not using Split K. Thus, this commit changes the splitk_batch_offset
parameter to a int32_t called k_size. This will avoid the constraint of
requiring a caller of MakeGemmTensorViews to use the SplitKBatchOffset
class while still providing the same functionality. Calls to
UniversalGemmKernel::MakeGemmTensorViews have been updated accordingly.

ecamartins committed:
StreamK Kernel RunGemm Implementation

Stream K cannot simply use UniversalGemmKernel's RunGemm for the
following reasons:

1. The UniversalGemmKernel::RunGemm function computes num_loop based on
   a static function of the TilePartitioner. That said, for Stream K,
num_loop must be computed using a member function (namely
GetCurrentIterLength from PR #2708).
2. The UniversalGemmKernel::RunGemm function requires the use of a
   SplitKBatchOffset object which is not used for Stream K since we are
not using Split K.

Thus, this change adds a RunGemm function in the StreamKKernel class.

arai713 committed:
initial implementation for operator() for StreamKKernel: adding stream-k algorithm and calls to RunGemm

ecamartins committed:
Fix indexing and offset issues for StreamK

These changes do the following:
- Ensure offsets along the M and N dimensions are multiplied by
  MPerblock or NPerBlock, respectively. This ensures tile window origins
are at the correct locations.
- Fix bug in the tile partitioner's GetTileIdxWithOffset. Now, we apply
  divmod to the given references to ensure correct values are available
to the caller.
- Added documentation in the Stream-K operator()

ecamartins committed:
Initial gtests for Stream-K

These changes add an initial gtest suite for the CK Tile Stream-K
kernel. Currently, due to bugs in the StreamKTilePartitioner (which will
be handled in a future PR), there are validation issues for certain
cases which may differ on different architectures. Thus, we opted to run
cases that are only fully data-parallel (skipping others). A guard was
added to Stream-K's IsSupportedArgument method to ensure that callers
are aware of this constraint. Additionally, to ensure testing
reproducibility, options for setting the number of CUs and occupancy
were added to MakeKernelArgs.

zjing14 commented:
@ecamartins Hi, it is great to see StreamK in CKTile. How about performance of StreamK vs SplitK for random shapes?

arai713 commented:
> @ecamartins Hi, it is great to see StreamK in CKTile. How about performance of StreamK vs SplitK for random shapes?

@zjing14 This is still the initial implementation, we have not had the chance to benchmark StreamK vs SplitK yet.

ecamartins committed:
Use GemmPipeline operator() variant that takes hot loop and tail num

In Stream-K, the num_loop value varies per WG and per iteration of a
Stream-K loop. So instead, we use the version of the GemmPipeline's
operator() function that takes in has_hot_loop and tail_num. This is
similar to what is done in Grouped GEMM.

CongMa13 commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -141,11 +141,17 @@ struct StreamKKernel
+    /// select their own to assist with test reproducibility, etc.
+    /// @return The kernel arguments for Stream-K.
+    CK_TILE_HOST static StreamKKernelArgs MakeKernelArgs(const StreamKHostArgs& host_args,
+                                                         int num_cu    = NumCU(),
+                                                         int occupancy = Occupancy())
```
Is that possible to make `num_cu` and `occupancy` in `StreamKHostArgs` so that we don't need to change the signature of `MakeKernelArgs`?

ecamartins replied:
We did consider that option; however, the `occupancy` is computed using the `hipOccupancyMaxActiveBlocksPerMultiprocessor` function. This function requires the kernel, which relies on the `TilePartitioner`, `GemmPipeline`, and `EpiloguePipeline`. So, if we moved the `occupancy` calculation to `StreamKHostArgs`, we would also need to know `TilePartitioner`, `GemmPipeline`, and `EpiloguePipeline`. So, we opted to include `num_cu` and `occupancy` in the `MakeKernelArgs` as it seems to be the least invasive option.

cgmillette commented on include/ck_tile/ops/gemm/kernel/gemm_tile_partitioner.hpp:
```diff
@@ -672,9 +672,7 @@ struct StreamKTilePartitioner
     CK_TILE_DEVICE void
     GetTileIdxWithOffset(uint32_t iter, uint32_t& tile_idx, uint32_t& iter_offset) const noexcept
     {
-        uint32_t tile_idx_val    = static_cast<uint32_t>(tile_idx);
```
Interesting.. So this function used to do nothing? Casts away the reference.

ecamartins replied:
In old CK, it had my current changes. I think this was missed when we converted between signed to unsigned in the StreamKTilePartitioner PR. This was just caught when I was testing the kernel, so we made the fix here.

cgmillette replied:
Is any other code using this? Might trigger some behaviour changes.

arai713 replied:
This function is used in the while loop in the operator function to determine the tile_idx

ecamartins replied:
Currently, we are the only ones using this code. This change was just meant to fix a bug that got introduced into the tile partitioner when we ported it from old CK.

cgmillette commented on test/ck_tile/gemm_streamk/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# Currently test_ck_tile_streamk is only built on gfx9
+if(GPU_TARGETS MATCHES "gfx9")
```
Eventually, we'd like to seamlessly target all arch's; can we add a ```#TODO: support all arches``` note

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -199,9 +263,78 @@ struct StreamKKernel
+
+        // An "iteration" denotes the multiplication of one macro tile in A with a macro tile in B.
+        // The total iteration length is the total of such multiplications performed.
+        uint32_t total_iter_length = iter_end - iter_start;
```
We may want to pull the readfirstlane up to here if the iter_end and iter_start will be the same for all threads. This might save a few vector ALU cycles from the partitioner's GetCurrentIterLength() call below. Then we can likely remove the other readfirstlane calls below because these should be scalar: total_iter_length, iter_end, iter_start.

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -199,9 +263,78 @@ struct StreamKKernel
+                    k_size);
+
+            // Prepare for next Stream-K loop iteration.
+            iter_end -= current_iter_length;
```
iter_start and iter_end are uint32_t, so this could potentially trigger some underflow issues. Should we consider using index_t for these values since we are counting down?

ecamartins replied:
Good point. We have altered the algorithm such that the WGs traverse macro tiles in the C tensor in the forward direction. As such now `iter_end` remains fixed and `iter_start` increments, which avoids triggering the underflow issues. See [7328bc7](https://github.com/ROCm/composable_kernel/pull/2781/commits/7328bc73eb927b2037053da41ac816d765d86857).

cgmillette commented on test/ck_tile/gemm_streamk/test_gemm_streamk_cases.inc:
```diff
@@ -0,0 +1,109 @@
+
+#pragma once
+
+#ifndef TEST_STREAM_K_CASES_INC
```
pragma once does the same thing here

arai713 committed:
changes from review: comments, move readfirstlane, remove ifndef

ecamartins committed:
Switch direction of C tensor traversal & add padding guard

Prior to this change, WGs travelled backwards through their assigned
macro tiles in the C tensor. For instance, if WG0 is responsible for C
tiles 0 and 1, it would first visit tile 1 then tile 0. This means that
the iter_end decrements in each iteration of the stream-K while loop.

Since we are working with unsigned integers, the subtraction operation
may not be safe. Thus, this change makes is such that WGs travel forward
so that their iter_start is incremented and their iter_end remains
fixed.

Additionally, we added a guard against WGs that are neither sk_blocks
nor dp_blocks to ensure such WGs do not participate in the GEMM.

Together, these changes make is such that the algorithm is correct when
sk_blocks is greater than zero.

cgmillette reviewed (dismissed):
LGTM

ecamartins committed:
Disable StreamK_M256_N256_K256_SKBlocks12 test case

This instance involves >=3 WGs contributing to each macro tile in C. Due
to the use of atomics, this is resulting in precision errors. These
errors will not persist once the reduction strategy is implemented. We
will re-enable this test then.

cgmillette reviewed (approved):
LGTM

vidyasagar-amd reviewed (approved):
LGTM

CongMa13 commented on test/ck_tile/gemm_streamk/test_gemm_streamk_util.hpp:
```diff
@@ -0,0 +1,282 @@
+                s, ck_tile::make_kernel<kBlockPerCu>(Kernel{}, grid_dims, block_dims, 0, kargs));
+        };
+
+        Run(ck_tile::integral_constant<ck_tile::memory_operation_enum,
```
We have to use atomic_add for both `StreamKReductionStrategy::Atomic` and `StreamKReductionStrategy::Reduction` ?

ecamartins replied:
Thanks for your comment! The `StreamKReductionStrategy::Atomic` will use atomic_add whereas the `StreamKReductionStrategy::Reduction` will use a set operation. The `StreamKReductionStrategy::Reduction` is currently not implemented, so there is a guard in `IsSupportedArgument` to prevent users from using  `StreamKReductionStrategy::Reduction`. We will need to update the tests and the `IsSupportedArgument` function once the reduction is implemented :)
